### ISSUE TITLE
ADC read in interrupt mode is not working properly. Need help

### ISSUE THREAD

**Issue by RhonyD** [cnlohr/ch32v003fun#474]
Here is my code:-
![image](https://github.com/user-attachments/assets/7da5de93-895f-4280-8154-36a97792c5de)

The problem is it's Showing the ADC value just after starting the program then it's showing the "ERROR" message in the while loop. What mistake am I doing? Need suggestion.

**Comment by TommyMurphyTM1234** [comment]
Doesn't your code have an inherent race condition with regard to `adcFlag` getting set to 1? Perhaps your `while` loop simply iterates before the interrupt handler has had a chance to set `adcFlag` to 1? If you're going to use this sort of handshaking between the main loop and an interrupt handler then you probably need to use a different approach and ideally atomic set/clear operations.

**Comment by RhonyD** [comment]
The adcFlag is set when conversion is complete and this flag is cleared in main loop. where is the problem I am unable to find the problem.

**Comment by TommyMurphyTM1234** [comment]
The setting and clearing of `adcFlag` is almost certainly not an atomic operation and there is nothing to enforce a critical section on access to flag by the two "threads" (main loop and interrupt handler).

- https://en.wikipedia.org/wiki/Critical_section

**Comment by maxgerhardt** [comment]
You need 
```cpp
volatile u8 adcFlag = 0;
volatile u16 ADC_VALUE = 0;
```
since these are variables that normal code and interrupt code shares. The compiler otherwise optimizes away reading the adcFlag multiple times.

https://docs.arduino.cc/language-reference/en/variables/variable-scope-qualifiers/volatile/

**Comment by cnlohr** [comment]
Also @RhonyD in the future, please do not use code-to-image conversions in github issues. It prevents search indexing and copy-pasting.

**Comment by RhonyD** [comment]
@maxgerhardt,@cnlohr,@TommyMurphyTM1234  thanks to all of you for your suggestions and guidance. By the way my issue is solved. I have tried adcFlag and ADC_VALUE with volatile key word and without it,and things worked perfectly. My final code is as follows:-

#include "ch32v003fun.h"
#include <stdio.h>
#include "GPIO_CONFIG.h"
#include "I2C.h"
#include "I2C_LCD.h"
#include "IT_CONFIG.h"
#include "ADC_CONFIG.h"

#define LED_PORT GPIO_PORT
#define LED_PIN 2

u8 adcFlag = 0;
u16 voltage = 0;

#if ADC_INJECTION
u16 ADC_VAL2 = 0, ADC_VAL0 = 0, ADC_VAL1 = 0;
#endif

#if ADC_POLL || ADC_INTERRUPT
u16 ADC_VAL = 0;
#define ADC_CHANNEL 1
#endif

void TIMER2_INIT(u16 psclr, u16 atlr)
{

	RCC->APB1PCENR |= (RCC_APB1Periph_TIM2);
	TIM2->CTLR1 |= TIM_CKD_DIV1;
	TIM2->PSC = psclr;
	TIM2->ATRLR = atlr;
	TIM2->SWEVGR |= TIM_UG;
	TIM2->DMAINTENR |= (TIM_IT_Update);
	TIM2->CTLR1 |= TIM_CEN;
}

void TIM2_IRQHandler()
{

	if (TIM2->INTFR & (TIM_IT_Update))
	{
		// TOGGLE_GPIO(LED_PIN);
		TIM2->INTFR &= ~(u16)TIM_IT_Update;
	}
}

#if ADC_INTERRUPT
void ADC1_IRQHandler()
{

	if (ADC1->CTLR1 & (1 << 5))
	{
		adcFlag = 1;
		ADC_VAL = ADC1->RDATAR;
	}
	/*
	if (ADC1->STATR& ADC_FLAG_EOC)
	{
		adcFlag = 1;
		ADC_VAL = ADC1->RDATAR;
	}
	*/
}
#endif

int main()
{
	SystemInit();
	GPIO_INIT(LED_PIN);
	TIMER2_INIT(28000 - 1, 999);
	i2c_init();
	Delay_Ms(1000);

#if ADC_INJECTION
	ADC_INIT(2, 2, 0, 0, 0);
#endif

#if ADC_POLL
	ADC_INIT(1);
#endif

#if ADC_INTERRUPT
	ADC_INIT(ADC_CHANNEL);

#endif

	GPIO_LOW(LED_PIN);
	Delay_Ms(1000);
	I2C_LCD_INIT(CURSOR_OFF);
	LCD_SEND_STRING("ADC test...");
	Delay_Ms(2000);
	LCD_CLEAR();
#if USE_TIMER2
//	NVIC_EnableIRQ(TIM2_IRQn);
#endif

#if USE_ADC_IT
	NVIC_EnableIRQ(ADC_IRQn);
#endif
	while (1)
	{
		TOGGLE_GPIO(LED_PIN);
		Delay_Us(50);
		LCD_CLEAR();
#if ADC_INJECTION
		GET_ADC_VALUE(&ADC_VAL0, &ADC_VAL1, &ADC_VAL2);
		LCD_SEND_INT_XY(0, 0, ADC_VAL1);
		LCD_SEND_STRING_XY(4, 0, "PA2");
		LCD_SEND_INT_XY(0, 1, ADC_VAL0);
		LCD_SEND_STRING_XY(5, 1, "PA1");
		LCD_SEND_INT_XY(9, 1, ADC_VAL2);
		LCD_SEND_STRING_XY(13, 1, "PC4");
#endif

#if ADC_POLL
		GET_ADC_VALUE(&ADC_VAL0);
		LCD_SEND_INT_XY(0, 0, ADC_VAL0);
#endif

#if ADC_INTERRUPT
		ADC1->CTLR2 |= ADC_SWSTART;
		Delay_Us(50);
		if (adcFlag)
		{
			adcFlag = 0;
			voltage = (ADC_VAL * 3300) / (1023);
			LCD_SEND_INT_XY(0, 0, voltage);
			LCD_SEND_STRING_XY(5, 0, "mv");
		}
		else
		{
			LCD_SEND_STRING_XY(0, 1, "Error..");
		}

#endif
		Delay_Ms(1000);
	}
}

**_This is my ADC_CONIFG.h:_**-

#ifndef _ADC_CONIFG_H
#define _ADC_CONIFG_H
#include "ch32v003fun.h"

#define ADC_INJECTION 0
#define ADC_POLL 0
#define ADC_INTERRUPT 1

#if ADC_INJECTION
void ADC_INIT(u8 ch, u8 a1, u8 a2, u8 a3, u8 a4);
void GET_ADC_VALUE(u16 *ADC_VAL0, u16 *ADC_VAL1, u16 *ADC_VAL2);
#endif

#if ADC_POLL
void ADC_INIT(u8 ch);
void GET_ADC_VALUE(u16 *ADC_VAL0);
#endif

#if ADC_INTERRUPT
void ADC_INIT(u8 ch);
#endif

#endif

**_This is my ADC_CONFIG.c:-_**

#include "ADC_CONFIG.h"

#if ADC_INJECTION

void ADC_INIT(u8 ch, u8 a1, u8 a2, u8 a3, u8 a4)
{

    // Enable ADC
    RCC->APB2PCENR |= RCC_APB2Periph_ADC1;

    // Reset the ADC to init all regs
    RCC->APB2PRSTR |= RCC_APB2Periph_ADC1;
    RCC->APB2PRSTR &= ~RCC_APB2Periph_ADC1;
    RCC->CFGR0 |= RCC_ADCPRE_0 | RCC_ADCPRE_1;
    //  Setup channel conversion situation
    /*
    ADC1->ISQR=ADC_JL_0 | ADC_JL_1 |     // 4 Total channels
        ADC_JSQ4_1 | ADC_JSQ4_0 | // Channel 3 (PD2)
        ADC_JSQ3_1 |              // Channel 2 (PC4)
        ADC_JSQ2_0 |              // Channel 1 (PA1)
        0;						  //Channel 0 (PA2)
    */

    ADC1->ISQR |= ((ch - 1) << 20);
    if ((ch == 1) && (a1 == 0))
    {
        ADC1->ISQR |= 0;
    }
    if ((ch == 1) && (a1 == 1))
    {
        // ADC1->ISQR |= ADC_JSQ2_0;
    }
    if ((ch == 2) && (a1 == 2))
    {
        ADC1->ISQR |= ADC_JSQ3_1;
    }
    if ((ch == 1) && (a1 == 3))
    {
    }

    // Setup sampling time
    // 0:7 => 3/9/15/30/43/57/73/241 cycles
    ADC1->SAMPTR2 |= ADC_SMP1_0 | ADC_SMP1_1 | ADC_SMP1_2;
    ADC1->SAMPTR2 |= ADC_SMP0_0 | ADC_SMP0_1 | ADC_SMP0_2;
    ADC1->SAMPTR2 |= ADC_SMP2_0 | ADC_SMP2_1 | ADC_SMP2_2;

    // turn on ADC and set rule group to sw trig
    // Additionally ADC_CONT says just continuously read the values.
    ADC1->CTLR2 = ADC_ADON | ADC_JEXTSEL | ADC_CONT;

    ADC1->CTLR1 = ADC_SCAN;

    // Reset calibration
    ADC1->CTLR2 |= ADC_RSTCAL;
    while (ADC1->CTLR2 & ADC_RSTCAL)
        ;

    // Calibrate
    ADC1->CTLR2 |= ADC_CAL;
    while (ADC1->CTLR2 & ADC_CAL)
        ;

    // Tell the ADC to start converting, continously.
    ADC1->CTLR2 |= ADC_JSWSTART;
}

void GET_ADC_VALUE(u16 *ADC_VAL0, u16 *ADC_VAL1, u16 *ADC_VAL2)
{
    *ADC_VAL2 = ADC1->IDATAR3; // channel 2 PC4
    *ADC_VAL1 = ADC1->IDATAR2; // channel 1 PA1
    *ADC_VAL0 = ADC1->IDATAR1; // channel 0 PA2
}

#endif

#if ADC_POLL
void ADC_INIT(u8 ch)
{

    // ADCCLK = 24 MHz => RCC_ADCPRE = 0: divide by 2
    RCC->CFGR0 &= ~(0x1F << 11);

    // Enable GPIOD and ADC
    RCC->APB2PCENR |= RCC_APB2Periph_GPIOD | RCC_APB2Periph_ADC1;

    // PD4 is analog input chl 7
    // GPIOD->CFGLR &= ~(0xf<<(4*4));	// CNF = 00: Analog, MODE = 00: Input

    // Reset the ADC to init all regs
    RCC->APB2PRSTR |= RCC_APB2Periph_ADC1;
    RCC->APB2PRSTR &= ~RCC_APB2Periph_ADC1;

    // Set up single conversion on chl 7
    ADC1->RSQR1 = 0;
    ADC1->RSQR2 = 0;
    ADC1->RSQR3 = ch; // 0-9 for 8 ext inputs and two internals

    // set sampling time for chl 7
    ADC1->SAMPTR2 &= ~(ADC_SMP0 << (3 * ch));
    ADC1->SAMPTR2 |= 7 << (3 * ch); // 0:7 => 3/9/15/30/43/57/73/241 cycles

    // turn on ADC and set rule group to sw trig
    ADC1->CTLR2 |= ADC_ADON | ADC_EXTSEL;

    // Reset calibration
    ADC1->CTLR2 |= ADC_RSTCAL;
    while (ADC1->CTLR2 & ADC_RSTCAL)
        ;

    // Calibrate
    ADC1->CTLR2 |= ADC_CAL;
    while (ADC1->CTLR2 & ADC_CAL)
        ;
}

void GET_ADC_VALUE(u16 *ADC_VAL0)
{
    // start sw conversion (auto clears)
    ADC1->CTLR2 |= ADC_SWSTART;

    // wait for conversion complete
    while (!(ADC1->STATR & ADC_EOC))
        ;

    // get result
    *ADC_VAL0 = ADC1->RDATAR;
}

#endif
#if ADC_INTERRUPT
void ADC_INIT(u8 ch)
{

    // ADCCLK = 24 MHz => RCC_ADCPRE = 0: divide by 2
    RCC->CFGR0 &= ~(0x1F << 11);

    // Enable GPIOD and ADC
    RCC->APB2PCENR &= ~RCC_APB2Periph_ADC1;
    RCC->APB2PCENR |= RCC_APB2Periph_ADC1;

    // PD4 is analog input chl 7
    // GPIOD->CFGLR &= ~(0xf<<(4*4));	// CNF = 00: Analog, MODE = 00: Input

    // Reset the ADC to init all regs
    //RCC->APB2PRSTR |= RCC_APB2Periph_ADC1;
    //RCC->APB2PRSTR &= ~RCC_APB2Periph_ADC1;

    // Set up single conversion on chl 7
    ADC1->RSQR1 = 0;
    ADC1->RSQR2 = 0;
    ADC1->RSQR3 = ch; // 0-9 for 8 ext inputs and two internals

    // set sampling time for chl 7
    ADC1->SAMPTR2 &= ~(ADC_SMP0 << (3 * ch));
    ADC1->SAMPTR2 |= 7 << (3 * ch); // 0:7 => 3/9/15/30/43/57/73/241 cycles

    // turn on ADC and set rule group to sw trig

    ADC1->CTLR2 |= ADC_ADON | ADC_EXTSEL;
    ADC1->CTLR1 |= ADC_IT_EOC;
    ADC1->CTLR2 |= ADC_SWSTART;

    // Reset calibration
    ADC1->CTLR2 |= ADC_RSTCAL;
    while (ADC1->CTLR2 & ADC_RSTCAL);
        

    // Calibrate
    ADC1->CTLR2 |= ADC_CAL;
    while (ADC1->CTLR2 & ADC_CAL);
        
}



#endif

**Comment by maxgerhardt** [comment]
Can you edit your last post to put that code in a tripple-backtick C code block? https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Comment by RhonyD** [comment]
'''
#include "ch32v003fun.h"
#include <stdio.h>
#include "GPIO_CONFIG.h"
#include "I2C.h"
#include "I2C_LCD.h"
#include "IT_CONFIG.h"
#include "ADC_CONFIG.h"

#define LED_PORT GPIO_PORT
#define LED_PIN 2

u8 adcFlag = 0;
u16 voltage = 0;

#if ADC_INJECTION
u16 ADC_VAL2 = 0, ADC_VAL0 = 0, ADC_VAL1 = 0;
#endif

#if ADC_POLL || ADC_INTERRUPT
u16 ADC_VAL = 0;
#define ADC_CHANNEL 1
#endif

void TIMER2_INIT(u16 psclr, u16 atlr)
{

	RCC->APB1PCENR |= (RCC_APB1Periph_TIM2);
	TIM2->CTLR1 |= TIM_CKD_DIV1;
	TIM2->PSC = psclr;
	TIM2->ATRLR = atlr;
	TIM2->SWEVGR |= TIM_UG;
	TIM2->DMAINTENR |= (TIM_IT_Update);
	TIM2->CTLR1 |= TIM_CEN;
}

void TIM2_IRQHandler()
{

	if (TIM2->INTFR & (TIM_IT_Update))
	{
		// TOGGLE_GPIO(LED_PIN);
		TIM2->INTFR &= ~(u16)TIM_IT_Update;
	}
}

#if ADC_INTERRUPT
void ADC1_IRQHandler()
{

	if (ADC1->CTLR1 & (1 << 5))
	{
		adcFlag = 1;
		ADC_VAL = ADC1->RDATAR;
		ADC1->CTLR2 |= ADC_SWSTART;
	}
	/*
	if (ADC1->STATR& ADC_FLAG_EOC)
	{
		adcFlag = 1;
		ADC_VAL = ADC1->RDATAR;
	}
	*/
}
#endif

int main()
{
	SystemInit();
	GPIO_INIT(LED_PIN);
	TIMER2_INIT(28000 - 1, 999);
	i2c_init();
	Delay_Ms(1000);

#if ADC_INJECTION
	ADC_INIT(2, 2, 0, 0, 0);
#endif

#if ADC_POLL
	ADC_INIT(1);
#endif



	GPIO_LOW(LED_PIN);
	Delay_Ms(1000);
	I2C_LCD_INIT(CURSOR_OFF);
	LCD_SEND_STRING("ADC test...");
	Delay_Ms(2000);
	LCD_CLEAR();
#if USE_TIMER2
//	NVIC_EnableIRQ(TIM2_IRQn);
#endif

#if USE_ADC_IT
	NVIC_EnableIRQ(ADC_IRQn);
#endif

u8 ch=0;


	while (1)
	{

#if ADC_INTERRUPT
	ADC_INIT(ch);

#endif

		TOGGLE_GPIO(LED_PIN);
		Delay_Us(50);
		LCD_CLEAR();
#if ADC_INJECTION
		GET_ADC_VALUE(&ADC_VAL0, &ADC_VAL1, &ADC_VAL2);
		LCD_SEND_INT_XY(0, 0, ADC_VAL1);
		LCD_SEND_STRING_XY(4, 0, "PA2");
		LCD_SEND_INT_XY(0, 1, ADC_VAL0);
		LCD_SEND_STRING_XY(5, 1, "PA1");
		LCD_SEND_INT_XY(9, 1, ADC_VAL2);
		LCD_SEND_STRING_XY(13, 1, "PC4");
#endif

#if ADC_POLL
		GET_ADC_VALUE(&ADC_VAL0);
		LCD_SEND_INT_XY(0, 0, ADC_VAL0);
#endif

#if ADC_INTERRUPT
		
		Delay_Us(50);
		if (adcFlag)
		{
			adcFlag = 0;
			voltage = (ADC_VAL * 3300) / (1023);
			LCD_SEND_INT_XY(0, 0, voltage);
			LCD_SEND_STRING_XY(5, 0, "mv");
		}
		else
		{
			LCD_SEND_STRING_XY(0, 1, "Error..");
		}
		ch++;
		if(ch>2)
		{
			ch=0;
		}

#endif
		Delay_Ms(1000);
	}
}
'''
'''
#ifndef _ADC_CONIFG_H
#define _ADC_CONIFG_H
#include "ch32v003fun.h"

#define ADC_INJECTION 0
#define ADC_POLL 0
#define ADC_INTERRUPT 1
#define CAL_VOL 75 // ADC_CALVOL_75PERCENT

#if ADC_INJECTION
void ADC_INIT(u8 ch, u8 a1, u8 a2, u8 a3, u8 a4);
void GET_ADC_VALUE(u16 *ADC_VAL0, u16 *ADC_VAL1, u16 *ADC_VAL2);
#endif

#if ADC_POLL
void ADC_INIT(u8 ch);
void GET_ADC_VALUE(u16 *ADC_VAL0);
#endif

#if ADC_INTERRUPT
void ADC_INIT(u8 ch);
#endif

#endif
'''
'''
#include "ADC_CONFIG.h"

#if ADC_INJECTION

void ADC_INIT(u8 ch, u8 a1, u8 a2, u8 a3, u8 a4)
{

    // Enable ADC
    RCC->APB2PCENR |= RCC_APB2Periph_ADC1;

    // Reset the ADC to init all regs
    RCC->APB2PRSTR |= RCC_APB2Periph_ADC1;
    RCC->APB2PRSTR &= ~RCC_APB2Periph_ADC1;
    RCC->CFGR0 |= RCC_ADCPRE_0 | RCC_ADCPRE_1;
    //  Setup channel conversion situation
    /*
    ADC1->ISQR=ADC_JL_0 | ADC_JL_1 |     // 4 Total channels
        ADC_JSQ4_1 | ADC_JSQ4_0 | // Channel 3 (PD2)
        ADC_JSQ3_1 |              // Channel 2 (PC4)
        ADC_JSQ2_0 |              // Channel 1 (PA1)
        0;						  //Channel 0 (PA2)
    */

    ADC1->ISQR |= ((ch - 1) << 20);
    if ((ch == 1) && (a1 == 0))
    {
        ADC1->ISQR |= 0;
    }
    if ((ch == 1) && (a1 == 1))
    {
        // ADC1->ISQR |= ADC_JSQ2_0;
    }
    if ((ch == 2) && (a1 == 2))
    {
        ADC1->ISQR |= ADC_JSQ3_1;
    }
    if ((ch == 1) && (a1 == 3))
    {
    }

    // Setup sampling time
    // 0:7 => 3/9/15/30/43/57/73/241 cycles
    ADC1->SAMPTR2 |= ADC_SMP1_0 | ADC_SMP1_1 | ADC_SMP1_2;
    ADC1->SAMPTR2 |= ADC_SMP0_0 | ADC_SMP0_1 | ADC_SMP0_2;
    ADC1->SAMPTR2 |= ADC_SMP2_0 | ADC_SMP2_1 | ADC_SMP2_2;

    // turn on ADC and set rule group to sw trig
    // Additionally ADC_CONT says just continuously read the values.
    ADC1->CTLR2 = ADC_ADON | ADC_JEXTSEL | ADC_CONT;

    ADC1->CTLR1 = ADC_SCAN;

    // Reset calibration
    ADC1->CTLR2 |= ADC_RSTCAL;
    while (ADC1->CTLR2 & ADC_RSTCAL)
        ;

    // Calibrate
    ADC1->CTLR2 |= ADC_CAL;
    while (ADC1->CTLR2 & ADC_CAL)
        ;

    // Tell the ADC to start converting, continously.
    ADC1->CTLR2 |= ADC_JSWSTART;
}

void GET_ADC_VALUE(u16 *ADC_VAL0, u16 *ADC_VAL1, u16 *ADC_VAL2)
{
    *ADC_VAL2 = ADC1->IDATAR3; // channel 2 PC4
    *ADC_VAL1 = ADC1->IDATAR2; // channel 1 PA1
    *ADC_VAL0 = ADC1->IDATAR1; // channel 0 PA2
}

#endif

#if ADC_POLL
void ADC_INIT(u8 ch)
{

    // ADCCLK = 24 MHz => RCC_ADCPRE = 0: divide by 2
    RCC->CFGR0 &= ~(0x1F << 11);

    // Enable GPIOD and ADC
    RCC->APB2PCENR |= RCC_APB2Periph_GPIOD | RCC_APB2Periph_ADC1;

    // PD4 is analog input chl 7
    // GPIOD->CFGLR &= ~(0xf<<(4*4));	// CNF = 00: Analog, MODE = 00: Input

    // Reset the ADC to init all regs
    RCC->APB2PRSTR |= RCC_APB2Periph_ADC1;
    RCC->APB2PRSTR &= ~RCC_APB2Periph_ADC1;

    // Set up single conversion on chl 7
    ADC1->RSQR1 = 0;
    ADC1->RSQR2 = 0;
    ADC1->RSQR3 = ch; // 0-9 for 8 ext inputs and two internals

    // set sampling time for chl 7
    ADC1->SAMPTR2 &= ~(ADC_SMP0 << (3 * ch));
    ADC1->SAMPTR2 |= 7 << (3 * ch); // 0:7 => 3/9/15/30/43/57/73/241 cycles

    // turn on ADC and set rule group to sw trig
    ADC1->CTLR2 |= ADC_ADON | ADC_EXTSEL;

    // Reset calibration
    ADC1->CTLR2 |= ADC_RSTCAL;
    while (ADC1->CTLR2 & ADC_RSTCAL)
        ;

    // Calibrate
    ADC1->CTLR2 |= ADC_CAL;
    while (ADC1->CTLR2 & ADC_CAL)
        ;
}

void GET_ADC_VALUE(u16 *ADC_VAL0)
{
    // start sw conversion (auto clears)
    ADC1->CTLR2 |= ADC_SWSTART;

    // wait for conversion complete
    while (!(ADC1->STATR & ADC_EOC))
        ;

    // get result
    *ADC_VAL0 = ADC1->RDATAR;
}

#endif
#if ADC_INTERRUPT
void ADC_INIT(u8 ch)
{

    // ADCCLK = 24 MHz => RCC_ADCPRE = 0: divide by 2
    RCC->CFGR0 &= ~(0x1F << 11);

    // Enable GPIOD and ADC
    RCC->APB2PCENR &= ~RCC_APB2Periph_ADC1;
    RCC->APB2PCENR |= RCC_APB2Periph_ADC1;

    // PD4 is analog input chl 7
    // GPIOD->CFGLR &= ~(0xf<<(4*4));	// CNF = 00: Analog, MODE = 00: Input

    // Reset the ADC to init all regs
    // RCC->APB2PRSTR |= RCC_APB2Periph_ADC1;
    // RCC->APB2PRSTR &= ~RCC_APB2Periph_ADC1;

    // Set up single conversion on chl 7
    ADC1->RSQR1 = 0;
    ADC1->RSQR2 = 0;
    ADC1->RSQR3 = ch; // 0-9 for 8 ext inputs and two internals

    // set sampling time for chl 7
    ADC1->SAMPTR2 &= ~(ADC_SMP0 << (3 * ch));
    ADC1->SAMPTR2 |= 7 << (3 * ch); // 0:7 => 3/9/15/30/43/57/73/241 cycles

    // turn on ADC and set rule group to sw trig

    ADC1->CTLR2 |= ADC_ADON | ADC_EXTSEL;
    ADC1->CTLR1 |= ADC_IT_EOC;
    ADC1->CTLR2 |= ADC_SWSTART;
    ADC1->CTLR1 &= ~(uint32_t)(3 << 25);
#if CAL_VOL == 50
    ADC1->CTLR1 |= ADC_CALVOL_50PERCENT;
#endif
#if CAL_VOL == 75
    ADC1->CTLR1 |= ADC_CALVOL_75PERCENT;
#endif
    // Reset calibration
    ADC1->CTLR2 |= ADC_RSTCAL;
    while (ADC1->CTLR2 & ADC_RSTCAL)
        ;

    // Calibrate
    ADC1->CTLR2 |= ADC_CAL;
    while (ADC1->CTLR2 & ADC_CAL)
        ;
}

#endif
'''

**Comment by RhonyD** [comment]
@[maxgerhardt](https://github.com/maxgerhardt) I have tried to do as per your instructions. Please see if this is what you wanted.

**Comment by maxgerhardt** [comment]
>and things worked perfectly. My final code is as follows:-

Okay, so the whole program is working? I'm still of the opinion that it's safer and correct to use `volatile` for variables shared with interrupts; It working may just be the compiler happening to emit the right code.

**Comment by RhonyD** [comment]
@maxgerhardt, yes you are right. This is just to inform you that the compiler is working fine. I will certainly use the volatile keyword. Thanks for your suggestion and guidance.

**Comment by TommyMurphyTM1234** [comment]
> > and things worked perfectly. My final code is as follows:-
> 
> Okay, so the whole program is working? I'm still of the opinion that it's safer and correct to use `volatile` for variables shared with interrupts; It working may just be the compiler happening to emit the right code.

I agree. If the code is compiled with optimisation enabled the behaviour may change. It may just be luck that the non-optimised version works right now. I also still think that the atomicity and critical section issues that I mentioned above should be addressed for similar reasons. As things stand the code doesn't look safe to me from a multiple "thread" point of view.

**Comment by RhonyD** [comment]
> > > and things worked perfectly. My final code is as follows:-
> > 
> > 
> > Okay, so the whole program is working? I'm still of the opinion that it's safer and correct to use `volatile` for variables shared with interrupts; It working may just be the compiler happening to emit the right code.
> 
> I agree. If the code is compiled with optimisation enabled the behaviour may change. It may just be luck that the non-optimised version works right now. I also still think that the atomicity and critical section issues that I mentioned above should be addressed for similar reasons. As things stand the code doesn't look safe to me from a multiple "thread" point of view. But none of these issues are specific to the ch32v003fun project/framework - they are general computer science/coding issues.

Yes Sir you are right. that's why I have already impelmented the volaite keyword in the code above. By the way thanks for taking time to read the issue and comment.
